File: packages/agent-sdk/src/canvas-kit.ts
/**
 * Canvas Kit — v0.5 / v0.6 block DSL for **Persona B hosted canvases**.
 *
 * These types describe the static blocks Renowide renders inside its own
 * buyer UI when you publish a Persona B agent (`renowide publish` with a
 * `renowide.yaml` that ships a `hire_page_canvas:` / `post_hire_canvas:`
 * tree directly in the manifest). The backend zod-validates the YAML
 * into this shape; your Python / TS code can construct the same blocks
 * at runtime (e.g. to render chat replies or dashboard tiles).
 *
 * ⚠️  Not the same protocol as Canvas Kit v2 (Path C — SDUI +
 * `custom_embed`). Canvas Kit v2 is a dynamic JSON response fetched
 * over HTTP from your backend on each render, with its own richer block
 * set, expression grammar, and HMAC signing. For that, import from
 * `@renowide/agent-sdk/canvas-kit-v2` (a re-export of `@renowide/types`)
 * or directly from `@renowide/types/canvas`.
 *
 * TL;DR:
 *   Persona B (this file)  → static blocks in `renowide.yaml` → hosted by Renowide
 *   Path C (canvas-kit-v2) → dynamic JSON per hire  → hosted by you, rendered by Renowide
 */

export interface BlockBase {
  when?: string;
}

export interface BlockHeader extends BlockBase {
  type: "header";
  text: string;
}
export interface BlockSection extends BlockBase {
  type: "section";
  text: string;
}
export interface BlockDivider extends BlockBase {
  type: "divider";
}
export interface BlockInfoCallout extends BlockBase {
  type: "info_callout";
  variant?: "info" | "warn" | "success";
  text: string;
}
export interface BlockImage extends BlockBase {
  type: "image";
  url: string;
  alt: string;
  caption?: string;
}
export interface BlockIntegrationButton extends BlockBase {
  type: "integration_button";
  provider: string;
  scopes?: string[];
  required?: boolean;
  label?: string;
}
export interface BlockApiKeyInput extends BlockBase {
  type: "api_key_input";
  id: string;
  label: string;
  placeholder?: string;
  help_url?: string;
  required?: boolean;
}
export interface BlockOAuthButton extends BlockBase {
  type: "oauth_button";
  provider: string;
  label?: string;
  scopes?: string[];
}
export interface BlockCheckbox extends BlockBase {
  type: "checkbox";
  id: string;
  text: string;
  required?: boolean;
  default?: boolean;
}
export interface BlockTextInput extends BlockBase {
  type: "text_input";
  id: string;
  label: string;
  placeholder?: string;
  required?: boolean;
  pattern?: string;
}
export interface BlockCTA extends BlockBase {
  type: "cta";
  text: string;
  action: string;
  style?: "primary" | "secondary";
}
export interface BlockLinkButton extends BlockBase {
  type: "link_button";
  text: string;
  url: string;
}
export interface BlockQuickReply extends BlockBase {
  type: "quick_reply";
  prompts: string[];
}
export interface BlockKPI extends BlockBase {
  type: "kpi";
  label: string;
  value: string;
  trend?: string;
}
export interface BlockTable extends BlockBase {
  type: "table";
  columns: string[];
  rows: Array<Array<string | number | boolean | null>>;
}

// ── v0.6 blocks ────────────────────────────────────────────────────────────

export interface BlockFileUpload extends BlockBase {
  type: "file_upload";
  id: string;
  label: string;
  required?: boolean;
  accept?: string[];
  max_mb?: number;
  help?: string;
}
export interface BlockDatePicker extends BlockBase {
  type: "date_picker";
  id: string;
  label: string;
  mode?: "date" | "datetime";
  required?: boolean;
  min?: string;
  max?: string;
  default?: string;
}
export interface BlockMarkdown extends BlockBase {
  type: "markdown";
  source: string;
}
export interface BlockCodeBlock extends BlockBase {
  type: "code_block";
  language?:
    | "plaintext" | "bash" | "sh" | "json" | "yaml" | "python"
    | "typescript" | "javascript" | "tsx" | "jsx" | "go" | "rust"
    | "sql" | "html" | "css" | "md";
  source: string;
  filename?: string;
}
export interface ChartSeries {
  label: string;
  data: number[];
}
export interface BlockChart extends BlockBase {
  type: "chart";
  chart_type?: "bar" | "line" | "pie" | "area";
  title?: string;
  labels: string[];
  series: ChartSeries[];
  stacked?: boolean;
}

export type CanvasBlock =
  | BlockHeader
  | BlockSection
  | BlockDivider
  | BlockInfoCallout
  | BlockImage
  | BlockIntegrationButton
  | BlockApiKeyInput
  | BlockOAuthButton
  | BlockCheckbox
  | BlockTextInput
  | BlockCTA
  | BlockLinkButton
  | BlockQuickReply
  | BlockKPI
  | BlockTable
  | BlockFileUpload
  | BlockDatePicker
  | BlockMarkdown
  | BlockCodeBlock
  | BlockChart;

// ── v0.6 Brand + Tool schema ───────────────────────────────────────────────

export interface Brand {
  primary_color?: string;
  accent_color?: string;
  text_color?: string;
  surface_color?: string;
  font_family?:
    | "inter"
    | "ibm_plex_sans"
    | "roboto"
    | "space_grotesk"
    | "source_serif_pro"
    | "jetbrains_mono"
    | "system";
  border_radius?: "none" | "small" | "medium" | "large";
}

export interface ToolInputField {
  name: string;
  type: "string" | "number" | "integer" | "boolean" | "date" | "file" | "enum";
  description?: string;
  required?: boolean;
  enum?: string[];
  default?: string | number | boolean;
}

export interface ToolManifest {
  name: string;
  display_name?: string;
  description: string;
  category: "read" | "write" | "communicate" | "analyse" | "act";
  inputs: ToolInputField[];
  requires_approval: boolean;
  icon?: string;
}

export interface CanvasVariant {
  id: string;
  weight?: number;
  blocks: CanvasBlock[];
}

export interface TileSource {
  type: "tool_call" | "static";
  tool?: string;
  data?: Record<string, unknown>;
}

export interface DashboardTile {
  id: string;
  title: string;
  size?: "small" | "medium" | "large";
  source?: TileSource;
  render: CanvasBlock[];
}

export interface ChatConfig {
  primary_color?: string;
  avatar?: string;
  greeting?: string;
  starter_prompts?: string[];
  canvas?: CanvasBlock[];
  variants?: CanvasVariant[];
}

export interface PostHireConfig {
  welcome_message?: string;
  welcome_canvas?: CanvasBlock[];
  variants?: CanvasVariant[];
}

export interface DashboardConfig {
  tiles: DashboardTile[];
}

// Convenience helpers for authors composing blocks programmatically.
export const header = (text: string, when?: string): BlockHeader => ({
  type: "header",
  text,
  when,
});
export const section = (text: string, when?: string): BlockSection => ({
  type: "section",
  text,
  when,
});
export const cta = (
  text: string,
  action: string,
  opts: { style?: "primary" | "secondary"; when?: string } = {},
): BlockCTA => ({
  type: "cta",
  text,
  action,
  style: opts.style ?? "primary",
  when: opts.when,
});
export const integrationButton = (
  provider: string,
  opts: { required?: boolean; scopes?: string[]; label?: string; when?: string } = {},
): BlockIntegrationButton => ({
  type: "integration_button",
  provider,
  required: opts.required ?? false,
  scopes: opts.scopes,
  label: opts.label,
  when: opts.when,
});

File: python/renowide-agent-sdk/src/renowide_agent_sdk/canvas_kit.py
"""
Hosted Layout v0.6 — block DSL for **Persona B hosted agents**.

Previously informally called "Persona B hosted canvas"; renamed to
Hosted Layout to eliminate the naming collision with Canvas Kit v2.
The protocol and schema are unchanged — only the name.

These TypedDicts describe the static blocks Renowide renders inside its
own buyer UI when you publish a Persona B agent (``renowide publish``
with a ``renowide.yaml`` that embeds a ``hire_page_canvas:`` /
``post_hire_canvas:`` tree directly in the manifest). They are the
Python twin of :mod:`@renowide/agent-sdk/canvas-kit` in TypeScript.

v0.6 additions: ``BlockFileUpload``, ``BlockDatePicker``, ``BlockMarkdown``,
``BlockCodeBlock``, ``BlockChart``, ``Brand``, ``ToolManifest``,
``CanvasVariant``.

.. warning::
   This is **not** the same protocol as Canvas Kit v2 (Path C — SDUI +
   ``custom_embed``). Canvas Kit v2 is a dynamic JSON response your
   backend returns per-hire, with a richer block set, expression grammar,
   and HMAC signing. For that, install the companion package::

       pip install renowide-canvas

   and import signing helpers / FastAPI router from
   :mod:`renowide_canvas`. This file stays for Persona B compatibility.
"""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional, TypedDict, Union


class _BlockBase(TypedDict, total=False):
    when: str


class BlockHeader(_BlockBase):
    type: Literal["header"]
    text: str


class BlockSection(_BlockBase):
    type: Literal["section"]
    text: str


class BlockDivider(_BlockBase):
    type: Literal["divider"]


class BlockInfoCallout(_BlockBase, total=False):
    type: Literal["info_callout"]
    variant: Literal["info", "warn", "success"]
    text: str


class BlockImage(_BlockBase, total=False):
    type: Literal["image"]
    url: str
    alt: str
    caption: str


class BlockIntegrationButton(_BlockBase, total=False):
    type: Literal["integration_button"]
    provider: str
    scopes: List[str]
    required: bool
    label: str


class BlockApiKeyInput(_BlockBase, total=False):
    type: Literal["api_key_input"]
    id: str
    label: str
    placeholder: str
    help_url: str
    required: bool


class BlockOAuthButton(_BlockBase, total=False):
    type: Literal["oauth_button"]
    provider: str
    label: str
    scopes: List[str]


class BlockCheckbox(_BlockBase, total=False):
    type: Literal["checkbox"]
    id: str
    text: str
    required: bool
    default: bool


class BlockTextInput(_BlockBase, total=False):
    type: Literal["text_input"]
    id: str
    label: str
    placeholder: str
    required: bool
    pattern: str


class BlockCTA(_BlockBase, total=False):
    type: Literal["cta"]
    text: str
    action: str
    style: Literal["primary", "secondary"]


class BlockLinkButton(_BlockBase, total=False):
    type: Literal["link_button"]
    text: str
    url: str


class BlockQuickReply(_BlockBase, total=False):
    type: Literal["quick_reply"]
    prompts: List[str]


class BlockKPI(_BlockBase, total=False):
    type: Literal["kpi"]
    label: str
    value: str
    trend: str


class BlockTable(_BlockBase, total=False):
    type: Literal["table"]
    columns: List[str]
    rows: List[List[Union[str, int, float, bool, None]]]


# ─── v0.6 blocks ─────────────────────────────────────────────────────────────


class BlockFileUpload(_BlockBase, total=False):
    type: Literal["file_upload"]
    id: str
    label: str
    required: bool
    accept: List[str]
    max_mb: int
    help: str


class BlockDatePicker(_BlockBase, total=False):
    type: Literal["date_picker"]
    id: str
    label: str
    mode: Literal["date", "datetime"]
    required: bool
    min: str
    max: str
    default: str


class BlockMarkdown(_BlockBase, total=False):
    type: Literal["markdown"]
    source: str


class BlockCodeBlock(_BlockBase, total=False):
    type: Literal["code_block"]
    language: str
    source: str
    filename: str


class ChartSeries(TypedDict, total=False):
    label: str
    data: List[float]


class BlockChart(_BlockBase, total=False):
    type: Literal["chart"]
    chart_type: Literal["bar", "line", "pie", "area"]
    title: str
    labels: List[str]
    series: List[ChartSeries]
    stacked: bool


CanvasBlock = Union[
    BlockHeader, BlockSection, BlockDivider, BlockInfoCallout, BlockImage,
    BlockIntegrationButton, BlockApiKeyInput, BlockOAuthButton, BlockCheckbox,
    BlockTextInput, BlockCTA, BlockLinkButton, BlockQuickReply, BlockKPI, BlockTable,
    BlockFileUpload, BlockDatePicker, BlockMarkdown, BlockCodeBlock, BlockChart,
]


# ─── v0.6 Brand + Tool schema ────────────────────────────────────────────────


class Brand(TypedDict, total=False):
    primary_color: str
    accent_color: str
    text_color: str
    surface_color: str
    font_family: Literal[
        "inter",
        "ibm_plex_sans",
        "roboto",
        "space_grotesk",
        "source_serif_pro",
        "jetbrains_mono",
        "system",
    ]
    border_radius: Literal["none", "small", "medium", "large"]


class ToolInputField(TypedDict, total=False):
    name: str
    type: Literal["string", "number", "integer", "boolean", "date", "file", "enum"]
    description: str
    required: bool
    enum: List[str]
    default: Union[str, int, float, bool]


class ToolManifest(TypedDict, total=False):
    name: str
    display_name: str
    description: str
    category: Literal["read", "write", "communicate", "analyse", "act"]
    inputs: List[ToolInputField]
    requires_approval: bool
    icon: str


class CanvasVariant(TypedDict, total=False):
    id: str
    weight: int
    blocks: List[CanvasBlock]


class TileSource(TypedDict, total=False):
    type: Literal["tool_call", "static"]
    tool: str
    data: Dict[str, Any]


class DashboardTile(TypedDict, total=False):
    id: str
    title: str
    size: Literal["small", "medium", "large"]
    source: TileSource
    render: List[CanvasBlock]


class ChatConfig(TypedDict, total=False):
    primary_color: str
    avatar: str
    greeting: str
    starter_prompts: List[str]
    canvas: List[CanvasBlock]
    variants: List[CanvasVariant]


class PostHireConfig(TypedDict, total=False):
    welcome_message: str
    welcome_canvas: List[CanvasBlock]
    variants: List[CanvasVariant]


# ─── Convenience constructors ────────────────────────────────────────────────


def header(text: str, when: Optional[str] = None) -> BlockHeader:
    out: BlockHeader = {"type": "header", "text": text}  # type: ignore[typeddict-item]
    if when:
        out["when"] = when
    return out


def section(text: str, when: Optional[str] = None) -> BlockSection:
    out: BlockSection = {"type": "section", "text": text}  # type: ignore[typeddict-item]
    if when:
        out["when"] = when
    return out


def cta(
    text: str,
    action: str,
    *,
    style: Literal["primary", "secondary"] = "primary",
    when: Optional[str] = None,
) -> BlockCTA:
    out: BlockCTA = {"type": "cta", "text": text, "action": action, "style": style}  # type: ignore[typeddict-item]
    if when:
        out["when"] = when
    return out


def integration_button(
    provider: str,
    *,
    required: bool = False,
    scopes: Optional[List[str]] = None,
    label: Optional[str] = None,
    when: Optional[str] = None,
) -> BlockIntegrationButton:
    out: BlockIntegrationButton = {  # type: ignore[typeddict-item]
        "type": "integration_button",
        "provider": provider,
        "required": required,
    }
    if scopes is not None:
        out["scopes"] = scopes
    if label is not None:
        out["label"] = label
    if when:
        out["when"] = when
    return out
